src/Adapter/TaxRulesGroup/CommandHandler/AddTaxRuleHandler.php

```diff
@@ -0,0 +1,175 @@
+        $behavior = $command->getBehaviorId()->getValue();
+        $description = $command->getDescription();
+        $taxRuleId = null;
+
+        $selectedCountries = $this->getCountryForTaxRule(
+            $this->countryDataProvider,
+            $this->langId,
+            $command->getCountryId()
+        );
+
+        $countriesWithErrors = [];
+        $stateIds = $command->getStateIds();
+
+        foreach ($selectedCountries as $country) {
+            $statesWithErrors = [];
+
+            /** @var StateId $stateId */
+            foreach ($stateIds as $stateId) {
+                $error = false;
+                try {
+                    $hasRuleForCountry = $this->assertUniqueBehaviorTaxRuleForCountry(
+                        $taxRulesGroupId,
+                        $country,
+                        $stateId->getValue()
+                    );
+
+                    if ($hasRuleForCountry) {
+                        $error = true;
+                    }
+
+                    $taxRule = new TaxRule();
+
+                    $taxRule->id_tax_rules_group = $taxRulesGroupId;
+                    $taxRule->behavior = $behavior;
+                    $taxRule->id_country = $country;
+                    $taxRule->id_tax = $taxId;
+                    $taxRule->id_state = $stateId->getValue();
+
+                    if (null !== $description) {
+                        $taxRule->description = $description;
+                    }
+
+                    if (null !== $command->getZipCode()) {
+                        $isInvalid = $this->assertIsValidZipCode(
+                            $zipCode,
+                            $country
+                        );
+
+                        if ($isInvalid) {
+                            $error = true;
+                        }
+                    }
+
+                    list($taxRule->zipcode_from, $taxRule->zipcode_to) = $taxRule->breakDownZipCode($zipCode);
+
+                    try {
+                        $this->validateTaxRuleFields($taxRule);
+                    } catch (TaxRuleConstraintException $e) {
+                        $error = true;
+                    }
+
+                    if (!$error && $taxRule->add()) {
+                        $taxRuleId = $taxRule->id;
+
+                        continue;
+                    }
+
+                    $error = true;
+                } catch (PrestaShopException $e) {
+                    $error = true;
+                }
+
+                if ($error) {
+                    $statesWithErrors[] = $stateId->getValue();
+                }
+            }
+
+            if (!empty($statesWithErrors)) {
+                $countriesWithErrors[$country] = $statesWithErrors;
+            }
+        }
+
+        if (!empty($countriesWithErrors)) {
+            throw new CannotAddTaxRuleException($countriesWithErrors);
+        }
+
+        //TODO find out how to handle hooks with multiple tax rule creation
```

Comment:
What was the strategy before ? In the legacy page ?

Reply:
Well if i understand correctly, Object model save and update methods have something like this `
// @hook actionObject*AddAfter

Hook::exec('actionObjectAddAfter', array('object' => $this));
Hook::exec('actionObject' . $this->getFullyQualifiedName() . 'AddAfter', array('object' => $this));` 

So on every object saving it executes hooks

Reply:
So it's fine.
You have called
```
$taxRule = new TaxRule();
...
$taxRule->add();
```
so the legacy hooks are still called, we have nothing else to do 😉